Add validation error to prevent checkout when there is no shipping method available

tarunvijwani opened the pull request:
With this PR, we're preventing checkout if no shipping methods are available. 

Fixes #8348

## Changes in the PR:
- Add the validation error in ShippingSelector to prevent checkout when no shipping rates are available.
- Add a notice to the checkout processor so that notice appears after clicking on the place order button.

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
|![image](https://user-images.githubusercontent.com/11503784/216732422-d263b199-9701-4174-981e-9f555d2bca6d.png)|![image](https://user-images.githubusercontent.com/11503784/221024165-bb27a4b5-a46b-4d4a-a39f-734a048900c7.png)|




### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Enable local pickup with pickup price.
2. Remove the shipping methods from the WC settings.
3. Go to the Checkout block and select the shipping option.
4. Confirm error notice is being displayed at the top.
5. Confirm you're not able to place the order.
6. Select local pickup.
7. Confirm you're able to place an order

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Add validation error to prevent checkout when there is no shipping method available

senadir commented on assets/js/base/components/cart-checkout/totals/shipping/index.tsx:
```diff
@@ -77,6 +81,17 @@ export const TotalsShipping = ( {
 		}
 	);
 
+	const { clearValidationError } = useDispatch( VALIDATION_STORE_KEY );
+	const { removeNotice } = useDispatch( 'core/notices' );
+
+	// Clear shipping rates validation error and notice when local pickup is selected.
+	useEffect( () => {
+		if ( prefersCollection ) {
+			clearValidationError( 'shipping-rates-error' );
+			removeNotice( 'shipping-rates-error', 'wc/checkout' );
+		}
+	}, [ clearValidationError, prefersCollection, removeNotice ] );
+
```
This doesn't seem like the right place to trigger this, it should be at the pickup/shipping selector level to ensure it only runs in Checkout and not both Cart and Checkout.

senadir commented on assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx:
```diff
@@ -50,6 +61,25 @@ const PackageRates = ( {
 		}
 	}, [ onSelectRate, rates, selectedOption ] );
 
+	// Set validation errors and display an error notice if there are no shipping rates available.
+	useEffect( () => {
+		if ( rates.length === 0 ) {
+			setValidationErrors( {
+				'shipping-rates-error': SHIPPING_RATE_ERROR,
+			} );
+			createErrorNotice(
+				decodeEntities(
+					'Shipping options are not available. Add a different address for shipping options'
+				),
+				{
+					id: 'shipping-rates-error',
+					context: 'wc/checkout',
+					isDismissible: true,
+				}
+			);
+		}
+	}, [ rates, setValidationErrors, createErrorNotice ] );
```
Shouldn't this error trigger at the shipping selector level? ideally we would only run into this scenario when we have both local pickup and regular shipping.

tarunvijwani replied:
Yes, thank you for the suggestion. I have moved the code to the Shipping selector.

senadir reviewed (changes requested):
~This is not working for me :/~

This is now working for me, but it still seems to be buggy.

If you go regular shipping, place, get an error, switch to local pickup, the error goes away, switch back to regular shipping, the error surfaces again for some seconds before it disappears.

senadir reviewed (changes requested):
Hey @tarunvijwani , we're going to put this on hold until @elizaan36 figures out some design next steps when it comes to validation and text.

elizaan36 commented:
When the user first lands on the Checkout block with no address provided, can we hide the "Shipping options" section?

Then if an address is entered that's not within a valid shipping zone, we can present a notice inline in the shipping options section. The top level notice isn't necessary in this case. Let me know what you think @tarunvijwani @senadir 

<img width="609" alt="image" src="https://user-images.githubusercontent.com/5121465/219547103-ed630ccb-d93b-4a45-964c-a2af6ba02e12.png">

senadir commented:
Hey @elizaan36 this will still get us to the same place of needing something to prevent the customer from placing order.

The issue here is that a shopper still has Local Pickup selected (even if not visible) and they can place the order with the error shown.

ralucaStan commented:
Noting that we don't have information-type notices implemented in the project. We could display a normal error notice in the shipping section until we have the new notice types added in the blocks. Does that sound good to you @elizaan36 ?

elizaan36 commented:
Yep, sounds good. Based on the conversation in Slack, there should be the same pattern/message for the same error whether local pickup is enabled or not.

When pickup is not enabled, there's an error notice in the Shipping options section. 
> There are no shipping options available. Please check your shipping address.

Then if user proceeds to checkout there's a top level notice.
> Sorry, this order requires a shipping option.

tarunvijwani commented:
Thank you, @senadir and @elizaan36, for the review. I have updated the code. 👍

mikejolley commented on assets/js/base/context/providers/cart-checkout/checkout-processor.ts:
```diff
@@ -157,8 +157,25 @@ const CheckoutProcessor = () => {
 	const checkValidation = useCallback( () => {
 		if ( hasValidationErrors() ) {
+			// If there is a shipping rates validation error, return the error message to be displayed.
+			if ( shippingRatesValidationError !== undefined ) {
```
Since this is within a callback and `shippingRatesValidationError` is not used anywhere else, could we move the lookup of this error inline?

e.g. `select( VALIDATION_STORE_KEY ).getValidationError( 'shipping-rates-error' ) !== undefined`

Might be cleaner?

tarunvijwani replied:
That's a great suggestion. I have updated the code.

mikejolley commented on assets/js/base/context/providers/cart-checkout/checkout-processor.ts:
```diff
@@ -157,8 +157,25 @@ const CheckoutProcessor = () => {
+			if ( shippingRatesValidationError !== undefined ) {
+				return {
+					errorMessage:
+						'Sorry, this order requires a shipping option.',
```
This needs to be localised.

tarunvijwani replied:
Ah missed localizing the strings. Thank you for bringing this up. I have added the localization.

mikejolley commented on assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx:
```diff
@@ -18,6 +21,11 @@ import { RatePrice, getLocalPickupPrices, getShippingPrices } from './shared';
 
+const SHIPPING_RATE_ERROR = {
+	hidden: true,
+	message: 'Shipping options are not available',
```
This needs to be localised.

mikejolley reviewed (changes requested):
Missing some localisation functions here, but it tested ok!

tarunvijwani commented on assets/js/data/utils/process-error-response.ts:
```diff
@@ -124,7 +124,7 @@ const processInvalidParamResponse = (
 export const processErrorResponse = (
 	response: ApiErrorResponse,
-	context: string | undefined
+	context?: string | undefined
```
This change will fix the TypeScript error present in the `assets/js/base/context/providers/cart-checkout/checkout-processor.ts` file for not passing [two arguments](https://github.com/woocommerce/woocommerce-blocks/blob/8932bc96ad8588e3feda2eba89014d1d19d92646/assets/js/base/context/providers/cart-checkout/checkout-processor.ts#L292).

tarunvijwani commented:
@mikejolley Thank you for the review. I have updated the code. 👍

mikejolley commented:
@tarunvijwani this needs a rebase - ready for final review?

Commit:
Prevent checkout when no shipping rates available

-  Add the validation error in ShippingSelector to prevent checkout when no shipping rates available.
- Add notice to Checkout processor so that notice appears after clicking on place order button.

Commit:
Fix TypeScript error in checkout shipping method block

-   Make multiple prop as optional in RatePrice component to fix TypeScript error.

tarunvijwani commented on assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/rate-price.tsx:
```diff
@@ -16,7 +16,7 @@ export const RatePrice = ( {
 	minRate: CartShippingPackageShippingRate | undefined;
 	maxRate: CartShippingPackageShippingRate | undefined;
-	multiple: boolean;
+	multiple?: boolean;
```
This change will fix the TypeScript error in `assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx` for not passing the `multiple` prop.

tarunvijwani commented:
> @tarunvijwani this needs a rebase - ready for final review?

@mikejolley I have fixed the conflicts. This is now ready for review. Thank you! 🙂

mikejolley reviewed (approved):
👍🏻

Commit:
Merge branch 'trunk' into add/8348-validation-error-to-prevent-checkout
